remove-hardwired-symbols-and-triggerItem

kayvank opened the pull request:
remove-hardwired-symbols, have not gotten to the the triggeritems yet.  All build and test issues resolved.

kayvank committed:
remove-hardwired-symbols-and-triggerItem

remove-hardwired-symbols-and-triggerItem

Mikolaj commented:
Cograts on making this compile. Could you try and remove most uses of `toContentSymbol`? In a separate commit, if you please. The goal of this exercise was to make the `ContentSymbol` abstract, so using `toContentSymbol` is a hack piercing through that abstarction, so should only be used if there's a good reason. The API to used instead of `toContentSymbol` is in `Game.LambdaHack.Content.RuleKind` (in which `toContentSymbol` is legit).

Mikolaj commented on GameDefinition/Content/ItemKind.hs:
```diff
@@ -17,7 +17,7 @@ import Game.LambdaHack.Core.Dice
 import Game.LambdaHack.Definition.Color
 import Game.LambdaHack.Definition.Defs
-import Game.LambdaHack.Definition.DefsInternal
+-- import Game.LambdaHack.Definition.DefsInternal
```
remove?

Mikolaj commented on GameDefinition/game-src/Client/UI/Content/Screen.hs:
```diff
@@ -45,8 +46,11 @@ standardLayoutAndFeatures = ScreenContent
-  , rapplyVerbMap = EM.fromList [('!', "quaff"), (',', "eat"), ('?', "read")]
-  , rFontFiles =
+  , rapplyVerbMap = EM.fromList [
+      (toContentSymbol '!', "quaff")
```
Could you insert a space between `(` and `t`?

Mikolaj commented on definition-src/Game/LambdaHack/Content/ItemKind.hs:
```diff
@@ -121,6 +120,7 @@ data ItemKind = ItemKind
   }
   deriving Show  -- No Eq and Ord to make extending logically sound
 
+
```
why the extra line?

Mikolaj commented on definition-src/Game/LambdaHack/Content/ModeKind.hs:
```diff
@@ -11,19 +11,16 @@ module Game.LambdaHack.Content.ModeKind
 import Game.LambdaHack.Core.Prelude
 
 import qualified Data.Text as T
-
```
Please keep the import blocks separate, as they were. Let me know if this requires explanation.

kayvank replied:
Thanks, I didn't know the reason. I looked it up in [kowainik.](https://kowainik.github.io/posts/2019-02-06-style-guide#imports)

Mikolaj commented on definition-src/Game/LambdaHack/Definition/Defs.hs:
```diff
@@ -1,9 +1,9 @@
+  ( GroupName(..), displayGroupName
   , ContentId, contentIdIndex
-  , ContentSymbol, displayContentSymbol
+  , ContentSymbol , displayContentSymbol, toContentSymbol
```
Once you remve the uses of `toContentSymbol` from the code, please remove the export of `toContentSymbol` here, too.

Mikolaj commented on definition-src/Game/LambdaHack/Definition/Defs.hs:
```diff
@@ -1,9 +1,9 @@
 -- | Basic types for content definitions.
 module Game.LambdaHack.Definition.Defs
-  ( GroupName, displayGroupName
+  ( GroupName(..), displayGroupName
```
Please don't expose the implementation of `GroupName`. If it's hard to just use the exposed functions, let's brainstorm.

Mikolaj replied:
Did you miss these, because they required "Load more"?

kayvank replied:
I did miss. I can use `displayGroupName`

Mikolaj replied:
Perfect.

Mikolaj commented on definition-src/Game/LambdaHack/Definition/Defs.hs:
```diff
@@ -14,7 +14,6 @@ module Game.LambdaHack.Definition.Defs
   ) where
 
 import Prelude ()
-
```
please keep the import blocks separated

Mikolaj replied:
Did you miss these, because they required "Load more"?

kayvank replied:
this was merge error on my end. will fix

Mikolaj commented on definition-src/Game/LambdaHack/Definition/DefsInternal.hs:
```diff
@@ -54,13 +54,14 @@ displayContentSymbol = id
 -- TODO2: extend to other content kinds than item kinds.
-{-
+
+
```
extra line

kayvank replied:
fixed

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/ActorUI.hs:
```diff
@@ -33,9 +35,14 @@ instance Binary ActorUI
+            -> (Bool, Bool, Bool, ContentSymbol ItemKind, Color.Color, ActorId)
 keySelected (aid, Actor{bhp, bwatch}, ActorUI{bsymbol, bcolor}) =
-  (bhp > 0, bwatch /= WSleep, bsymbol /= '@', bsymbol, bcolor, aid)
+  (bhp > 0
```
space

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/ActorUI.hs:
```diff
@@ -52,8 +59,10 @@ tryFindHeroK :: ActorDictUI -> FactionId -> Int -> State
 tryFindHeroK d fid k s =
   let c | k == 0          = '@'
-        | k > 0 && k < 10 = Char.intToDigit k
+        | k > 0 && k < 10 = Char.intToDigit $ k
```
the $ is spurious

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/ActorUI.hs:
```diff
@@ -52,8 +59,10 @@ tryFindHeroK :: ActorDictUI -> FactionId -> Int -> State
+        | k > 0 && k < 10 = Char.intToDigit $ k
         | otherwise       = ' '  -- no hero with such symbol
+      theSymbol = toContentSymbol c
+
```
extra line

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/DrawM.hs:
```diff
@@ -331,14 +331,14 @@ drawFrameActor drawnLevelId = do
               bg = Color.HighlightNone
               fg = bcolor
-         in Color.attrCharToW32 $ Color.AttrChar Color.Attr{..} bsymbol
+         in Color.attrCharToW32 $ Color.AttrChar Color.Attr{..} (DefsInternal.displayContentSymbol bsymbol)
```
This one can be gotten from `Defs`, so it's better not to refer to the internal one.

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Watch/WatchSfxAtomicM.hs:
```diff
@@ -61,6 +62,7 @@ import qualified Game.LambdaHack.Definition.Ability as Ability
 import qualified Game.LambdaHack.Definition.Color as Color
 import           Game.LambdaHack.Definition.Defs
 
+
```
extra line

kayvank committed:
remove-hardwired-symbols-and-triggerItem

Issue #152, removing uses of toContentSymbol where possible

Mikolaj commented:
The failed CI seem to be doctests. I guess they just need updating.

Mikolaj commented:
README tells how to run doctests. It may be easier than finding it in the CI script.

kayvank committed:
remove-triggerItem

issue 152, remove all hardwired item symbols as well as of TriggerItem.
This commit removes the `flingTs` function before continuing to removing
the trigger item. `flingTs`  added text to `tiverb` and `tiobject`. For now
this commit hardwired these texts in the
`Game.LambdaHack.Client.UI.Content.Input` function `descIs`

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Content/Input.hs:
```diff
@@ -6,7 +6,7 @@ module Game.LambdaHack.Client.UI.Content.Input
   , mouseLMB, mouseMMB, mouseMMBMute, mouseRMB
   , goToCmd, runToAllCmd, autoexploreCmd, autoexplore25Cmd
-  , aimFlingCmd, projectI, projectA, flingTs, applyIK, applyI
+  , aimFlingCmd, projectI, projectA,  applyIK, applyI
```
space

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Content/Input.hs:
```diff
@@ -212,13 +212,13 @@ autoexplore25Cmd :: HumanCmd
-projectICmd :: [TriggerItem] -> HumanCmd
-projectICmd ts = ComposeUnlessError (ChooseItemProject ts) Project
+projectICmd :: HumanCmd
+projectICmd  = ComposeUnlessError (ChooseItemProject []) Project --
```
space

Mikolaj replied:
and empty comment

Mikolaj replied:
and I'd hope you can remove the argument from `ChooseItemProject`

Mikolaj commented on GameDefinition/game-src/Client/UI/Content/Input.hs:
```diff
@@ -75,10 +75,10 @@ standardKeysAndMouse = InputContentRaw $ map evalKeyDef $
   , ("comma", grabItems "")  -- only show extra key, not extra entry
   , ("r", dropItems "remove item(s)")
-  , ("f", addCmdCategory CmdItemMenu $ projectA flingTs)
+  , ("f", addCmdCategory CmdItemMenu $ projectA [])
```
I'd hope you can remove the argument from `projectA`

Mikolaj commented on GameDefinition/game-src/Client/UI/Content/Input.hs:
```diff
@@ -75,10 +75,10 @@ standardKeysAndMouse = InputContentRaw $ map evalKeyDef $
   , ("C-f", addCmdCategory CmdItemMenu
             $ replaceDesc "auto-fling and keep choice"
-            $ projectI flingTs)
+            $ projectI [])
```
I'd hope you can remove the argument altogether from projectI --- the whole list argument

Mikolaj replied:
just as you removed it from `projectICmd`

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Content/Input.hs:
```diff
@@ -156,7 +156,7 @@ mouseLMB goToOrRunTo desc =
     , (CaHPGauge, Macro ["KP_Begin", "C-v"])
     , (CaHPValue, Wait)
-    , (CaLeaderDesc, projectICmd flingTs) ]
+    , (CaLeaderDesc, projectICmd ) ]
```
space

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Content/Input.hs:
```diff
@@ -212,13 +212,13 @@ autoexplore25Cmd :: HumanCmd
 
 projectI :: [TriggerItem] -> CmdTriple
-projectI ts = ([CmdItem], descIs ts, projectICmd ts)
+projectI ts = ([CmdItem], descIs ts, projectICmd )
```
space

Mikolaj commented on engine-src/Game/LambdaHack/Client/UI/Content/Input.hs:
```diff
@@ -263,7 +252,7 @@ dropItems t = ([CmdItemMenu, CmdItem], t, dropCmd)
 descIs :: [TriggerItem] -> Text
 descIs [] = "trigger an item"
-descIs (t : _) = makePhrase [tiverb t, tiobject t]
+descIs _ = makePhrase ["fling", "in-range projectile"]
```
I'd rather inline the whole `makePhrase ["fling", "in-range projectile"]` into `projectI`, because `descIs` is also used in `applyIK` so we can't change it freely.
